parlor.go

```diff
@@ -6,39 +6,44 @@ import (
 	"github.com/stretchr/testify/suite"
 )
 
-type Parlor struct {
-	suite.Suite
+type SetupSubtestSuite interface {
+	SetupSubtest()
+}
+
+type TearDownSubtestSuite interface {
+	TearDownSubtest()
 }
 
 type TestingParlor interface {
 	suite.TestingSuite
-	suite.SetupTestSuite
-	suite.TearDownTestSuite
+	SetTestingParlor(TestingParlor)
+}
+
+type Parlor struct {
+	suite.Suite
+	testingParlor TestingParlor
+}
+
+func (p *Parlor) SetTestingParlor(tp TestingParlor) {
+	p.testingParlor = tp
 }
 
 func Run(t *testing.T, parlor TestingParlor) {
+	parlor.SetTestingParlor(parlor)
 	suite.Run(t, parlor)
 }
 
-func (p *Parlor) Run(
-	name string,
-	subtest func(),
-	tp TestingParlor,
-) bool {
-	return p.RunWithSetupAndTeardown(
-		name,
-		subtest,
-		tp.SetupTest,
-		tp.TearDownTest,
-	)
-}
+func (p *Parlor) Run(name string, subtest func()) bool {
+	setup := func() {}
+	if i, ok := p.testingParlor.(SetupSubtestSuite); ok {
```

Comment:
Cool refactor that removes the need for separate `Run` and `RunWithSetupAndTeardown` funcs

Reply:
This took me a minute to understand. If I'm getting it right, the object we pass in when we call `SetTestingParlor(tp TestingParlor)` can implement the `SetupSubtestSuite` or `TearDownSubtestSuite` methods, but don't need to, since they are not required to. The only required interface is a TestingParlor interface, which embeds the `suite.Suite` interface and requires the `SetTestingParlor` method. So here we check to see if those interfaces have been implemented, otherwise use a no-op func for them.